### ISSUE TITLE
Replace Browserify?

### ISSUE THREAD

**Issue by jhilde** [studentinsights/studentinsights#1161]
In our team meeting on 10/13, a couple of us were complaining about the very long build times browserify takes as we're developing javascript. So we decided to look a bit at webpack.

Turns out Rails is moving toward a more modern understanding of the place of javascript on the front end. [Evidenced by a javascript folder now at the app level, not relegated to being in the app/assets]. Rails now provides the [webpacker gem](https://github.com/rails/webpacker) to facilitate webpack in rails apps. It appears that as of Rails 5.1, this is the recommended method to building and packaging javascript assets. 

The dev experience includes a wrapper around webpack-dev-server that allows hot recompiling of javascript. It also includes all the goodness of webpack in terms of code splitting and dependency mapping. (I haven't tried it with the complete app but the speed looks promising. I often get caught in Browserify hell. Webpack at least shows you progress unlike browserify).

From a first look, my short term recommendation is that we create one large entry point (bundle) for all the application code. This will be the easiest migration path because of the way our javascript is currently setup.

As we migrate to React 16 and hopefully move to using import rather than window.shared, it may make sense to create an entry point for the vendor javascript that is common to most of our pages and perhaps an entry point for each page. 

@alexsoble and @kevinrobinson, have you used webpacker in other projects? Any thoughts on stragegy?

**Comment by alexsoble** [comment]
Yes I use webpacker in a Rails 5.1 project! 👍

**Comment by alexsoble** [comment]
@jhilde In terms of prioritizing code-quality features, would you rank this one as more important or #1121?

**Comment by kevinrobinson** [comment]
Sounds good!  Here's what we learned trying to do this last time: https://github.com/studentinsights/studentinsights/pull/861#issuecomment-284084767, which might be outdated, or we might find other ways of tackling it.

It'd be awesome if this is a drop-in replacement, but my assumption is that it won't be and that iterating on the production config will require testing in the demo or staging site.  If not, then building towards JS-centric tooling that's separate from Rails seems like a better long-term path (eg `create-react-app`).

**Comment by jhilde** [comment]
Note to self, search the issues. ;-)  I recalled there being some conversation but wasn't up to speed on webpack, etc. 

I'm a big fan of create-react-app backed with a restish API.

So yes, if we all agree that's the future path, might be better time spent looking at that. Would be interesting to add some json endpoints to studentinsights and spin up a little create-react-app. I use create-react-app backed by AWS lambda functions for my little apps ([github.com/freedomconnexion/registration-web](http://github.com/freedomconnexion/registration-web)). I've been playing around with using jwt for endpoint authentication...it's pretty straightforward in rails.

**Comment by jhilde** [comment]
And yes @alexsoble, now that you've done some great iterating on the structure of the importers, I'd love to dive into detailed logging in import_record. 

(At least I don't have to wait for browserify for the work on the importers!)

**Comment by jhilde** [comment]
@kevinrobinson @alexsoble So should I close, modify the title, or leave open for now?

**Comment by alexsoble** [comment]
@kevinrobinson Curious to hear more of your thinking on the comparison between `create-react-app` and `webpacker` — what are the major pros of each approach in your view? 

@jhilde Why don't we make the title "Replace Browserify" for now?

**Comment by alexsoble** [comment]
A couple of other things I'm wondering about:

+ Will we need to migrate completely from `window.shared` to `import`/`export` before we make this jump?
+ Will we need to replace `main.jsx` files with a single entry point and React Routing before we make this jump? 
+ Are either of these pieces we can tackle in a separate PR to prepare for moving to `create-react-app` or `webpacker`?

**Comment by jhilde** [comment]
Hey @alexsoble....

create-react-app uses webpacker under the covers. Migrating to create-react-app is more about separation of the front end from the back end. 

My big picture thought is that we may or may not migrate at this time, but to the point @sentientmachine (I think) made last week, we should have a design doc on new components going forward. My list would include:

- use the prop-types library
- extend React.Component vs React.createClass
- favor imports vs. window.shared when possible

And then, upgrading to React 16 when we are ready. We'll need to do this anyways eventually.

As a separate item, we could create a proof of concept with create-react-app that:

- Adds token authentication to our rails app
- Adds a json endpoint for a dataset (could just do a current page's serializedData for a first pass)
- Creates a basic create-react-app that allows the user to authenticate, building a simple page (StudentProfile?) from data retrieved through the new json endpoint?

From there, we can decide:

- Is it worth it?
- How best to structure it with an eye toward as much code reuse as possible?
- How we would migrate?

Does that sound reasonable?

**Comment by alexsoble** [comment]
Hey @jhilde!

Did some reviewing past branches in this direction with @kevinrobinson. One thing we came up with was a list of dependencies that we'll need to change before we can move to an option like webpacker. Here are some notes so that we can break this into incremental pieces: 

### Things to change before we move to webpacker

+ we'll need a single point of entry, we don't use modules everywhere yet
+ we'll need to change the JS tests and the teaspoon configuration 
+ we'll need to look at production deploys

### Things to change before we move to create-react-app in particular
+ json endpoints
+ authentication mechanism (something like jwt, like you mentioned)
+ migration strategy for SPA

The single point of entry/ES6 modules jumps out at me as the first thing we could tackle. This is a prerequisite to moving off browserify-rails. It's also a change that we could ship and deploy in its own PR, celebrate and have a little online party for, and then return to the big-picture goal of moving away from browserify-rails with one less obstacle in our way. I'll split it off into its own issue and tie it to this PR. This will involve changing lots of JS but with the power of our team will clear it faster than we think and it will feel great when we land!

**Comment by alexsoble** [comment]
@jhilde See #1180 as a starting point!

**Comment by kevinrobinson** [comment]
Closed by https://github.com/studentinsights/studentinsights/pull/1182
